### DISCUSSION TITLE
how do I create a custom get_candidates function?

### DISCUSSION THREAD

**Discussion by moyid** [explosion/spaCy#7151] [Help: Coding & Implementations] [answered]
I'm trying to register a custom get_candidates function for my entity linking pipeline. I keep getting the error that it is not Callable. I have at this point swapped out my custom function with the default get_candidates from spacy, and I'm still getting the error, so the issue must be with my setup. -- apologies because I'm not familiar with setting up with the config files. Please help!

Here is my setup:

I have a scripts/custom_functions.py file, where I have:

```
import spacy
from spacy.kb import Candidate, KnowledgeBase, get_candidates

@spacy.registry.misc("CustomGetCandidates.v1")
def custom_create_candidates()  -> Callable[[KnowledgeBase, "Span"], Iterable[Candidate]]:
	return get_candidates
```
^ so I'm using the same default get_candidates, I'm just trying to see how to set it up so that I can write my own get_candidates function. 

And then in my nel.cfg file I have:

```
[components.entity_linker]
factory = "entity_linker"
entity_vector_length = 64
get_candidates = "CustomGetCandidates.v1"
incl_context = true
incl_prior = true
labels_discard = []

[components.entity_linker.model]
@architectures = "spacy.EntityLinker.v1"
nO = null
```

now when I run `spacy project run train`  I get the error:

```
✘ Config validation error
entity_linker -> get_candidates   CustomGetCandidates.v1 is not callable
```

PS - the set up is edited based on the nel_emerson example here: https://github.com/explosion/projects/tree/v3/tutorials/nel_emerson

**Comment by svlandeg** [comment]
Hi! This is a good question for the discussion forum, so I'll move it there. You may get a notification that your issue was closed/locked, but the thread is still open at the forum and you should be forwarded their automatically :-)

**Comment by moyid** [comment]
Thank you for moving this to discussions. Hoping to receive some feedback soon!

**Comment by moyid** [comment, marked as answer]
I have figured out what's wrong, I just need to change the config a little bit:

```
[components.entity_linker]
factory = "entity_linker"
entity_vector_length = 64
get_candidates = {"@misc": "CustomGetCandidates.v1"}
incl_context = true
incl_prior = true
labels_discard = []

[components.entity_linker.model]
@architectures = "spacy.EntityLinker.v1"
nO = null
```

**Reply by svlandeg** [reply]
Great, thanks for reporting back!
